README.md
# rapflag
Bitfinex and Poloniex Exporter for your Taxman.

### example installation instructions
```
yum install ruby-devel gcc-c++ make
gem install public_suffix:2.0.5 rapflag
```

### set your keys, desired currencies in
```
 ./etc/config.yml
```
sample
```
currencies: [ 'USD', 'BTC', 'BFX']
api_key: ''
secret: ''
poloniex_api_key: ''
poloniex_secret: ''
```
comment out "currencies" if you want to export all currencies.
### run as follows
```
Options:
  -c, --clean       Create Bitfinex summary of transactions by day
  -p, --poloniex    Use Poloniex API instead of Bitfinex API
  -d, --dump        Use Poloniex API and dump history into CSV files
  -h, --help        Show this message
```
Enjoy your reports. More to come. Feedback always welcome.

### License
GPLv3.0, see License File.


History.md
# 1.0.7 of 2017.11.08

* Add LTC and BCH to list of default currencies
* ignore entries like Bitcoin Gold snapshot step3 on Funding wallet from income

# 1.0.6 of 2017.09.27

* Fix calculation of daily income/balance

# 1.0.5 of 2017.09.27

* Fix emitting totals for each day

# 1.0.4 of 2017.09.26

* Consider dividend and distribution as income

# 1.0.3 of 2017.07.26

* <ticker>_total has balance + sum of balance fields of [deposit,trading,exchange]
* Added spec test for BTC. Checks balance of 2017.05.23

# 1.0.2 of 2017.07.26

* Correct calculation of balance for total files. 
* Consider settlement as income
* Correct calculation of balance for total files. Consider settlement (for dump)
* Remove empty test target
* Implemented dump_history also for bitfinex


# 1.0.1 of 2017.07.26

* Create correct entries for intermediate dates

# 1.0.0 of 2017.05.28

* Handle two transactions with same timestamp
* Fix duplicated income
* Do not emit rate and balance_in_usd for bitfinex
* Use ',' as column separator.

# 0.0.9 of 2017.05.24

* Add <currency>_total files for bitfinex

# 0.0.8 of 2017.04.17

* Wait (possibly several times) when we hit the rate limiter of the bitfinex REST API while fetching the currency rates

# 0.0.7 of 2017.04.17

* Wait (possibly several times) when we hit the rate limiter of the bitfinex REST API while fetching the histor

# 0.0.6 of 2017.04.11

* Download complete lending history not only he default of 500 items

# 0.0.5 of 2017.04.10

* Fixed daily balance for poloniex

# 0.0.4 of 2017.04.05

* Added options -p and -d for Poloniex
* Fixed problem with inexisting spec/data/*.json files

# 0.0.3 of 2017.04.02

* Output files start with the currency name




spec/spec_helper.rb
begin
require 'pry'
rescue LOAD_ERROR
end

require 'rapflag/version'
require 'rapflag/config'
require 'bitfinex-api-rb'
require 'poloniex'

RAPFLAG::Config['poloniex_api_key'] = 'dummy poloniex_api_key'
RAPFLAG::Config['poloniex_secret']  = 'dummy poloniex_secret'
RAPFLAG::Config['api_key'] = 'dummy bitfinex api_key'
RAPFLAG::Config['secret']  = 'dummy bitfinex secret'

module RAPFLAG
  @@outputDir = File.expand_path(File.join(File.dirname(__FILE__),  'output'))
end

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/rapflag_spec.rb
# encoding: utf-8
require 'spec_helper'
require 'rapflag/fetch'
require 'vcr'

VCR.configure do |config|
  config.cassette_library_dir = "fixtures/vcr_cassettes"
  config.hook_into :faraday # or :fakeweb
end

CSV_Test_File = File.expand_path(File.join(__FILE__, '..', '..', 'output/BTC_exchange.csv'))
SUMMARY_EXCHANGE_BTC_File = File.expand_path(File.join(__FILE__, '..', '..', 'output/BTC_exchange_summary.csv'))
SUMMARY_DEPOSIT_BFX_File = File.expand_path(File.join(__FILE__, '..', '..', 'output/BFX_deposit_summary.csv'))

VCR.eject_cassette # we use insert/eject around each example
describe RAPFLAG do
  # include ServerMockHelper
  before(:all) do
  end
  after(:all) do
  end
  context 'bitfinex' do
    before(:all) do
      VCR.use_cassette("rapflag", :record => :new_episodes) do
        FileUtils.rm_f(CSV_Test_File) if File.exist?(CSV_Test_File)
        FileUtils.rm_f(SUMMARY_DEPOSIT_BFX_File) if File.exist?(SUMMARY_DEPOSIT_BFX_File)
        FileUtils.rm_f(SUMMARY_EXCHANGE_BTC_File) if File.exist?(CSV_Test_File)
        expect(File.exist?(CSV_Test_File)).to eql(false)
        @rap = RAPFLAG::History.new('exchange', 'BTC')
        @rap.fetch_csv_history
        @rap.create_csv_file
      end
    end
    context 'history' do
      it 'should have correct currency' do
        expect(@rap.currency).to eql('BTC')
      end
      it 'should have correct size' do
        expect(@rap.history.size).to eql(206)
      end
    end
  end
  context 'bitfinex CSV' do
    context 'csv' do
      it 'should have generated a correct CSV file' do
        expect(File.exist?(CSV_Test_File)).to eql(true)
        lines = IO.readlines(CSV_Test_File)
        expect(lines.first.chomp).to eql('currency,amount,balance,description,date_time')
        expect(lines[1].chomp).to eql(
          'BTC,-0.00000005,0.0,Transfer of 0.0 BTC from wallet Exchange to Deposit on wallet Exchange,2016.12.03 21:20:47')
      end
    end
  end
  context 'exchange option --clean' do
    before(:all) do
      @date_bfx_1 = Date.new(2017,1,10)
      @date_btx_1 = Date.new(2017,1,21)
      @date_btx_2 = Date.new(2017,1,10)
      VCR.use_cassette("rapflag", :record => :new_episodes) do
        FileUtils.rm_f(SUMMARY_EXCHANGE_BTC_File) if File.exist?(CSV_Test_File)
        expect(File.exist?(SUMMARY_EXCHANGE_BTC_File)).to eql(false)
        @exchange = RAPFLAG::History.new('exchange', 'BTC')
        @exchange.fetch_csv_history
        @exchange.create_summary
        @bfx   = @exchange.get_usd_exchange(@date_bfx_1, 'BFX')
        @btx_1 = @exchange.get_usd_exchange(@date_btx_1, 'BTC')
        @btx_2 = @exchange.get_usd_exchange(@date_btx_2, 'BTC')
      end
    end
    it 'should have generated a correct summary CSV file' do
      expect(File.exist?(SUMMARY_EXCHANGE_BTC_File)).to eql(true)
      lines = IO.readlines(SUMMARY_EXCHANGE_BTC_File)
      expect(lines.first.chomp).to eql('currency,date,income,balance,rate,balance_in_usd')
      expect(lines[1].chomp).to eql('BTC,2016.01.15,0.0,8.99788147,,')
      expect(lines[-1].chomp).to eql('BTC,2016.12.03,0.0,0.0,765.46,0.0')
    end
    it 'should have NOT have generated a correct summary deposit BFX CSV file' do
      expect(File.exist?(SUMMARY_DEPOSIT_BFX_File)).to eql(false)
    end
    it 'should have the correct BTC -> USD rate' do
      expect(@btx_1).to eql 924.02
      expect(@btx_2).to eql 905.76
    end
    it 'should have the correct BFX -> USD rate' do
      expect(@bfx).to eql 0.5697
    end
  end
  context 'deposit option --clean' do
    before(:all) do
      FileUtils.rm_f(SUMMARY_EXCHANGE_BTC_File) if File.exist?(SUMMARY_EXCHANGE_BTC_File)
      FileUtils.rm_f(SUMMARY_DEPOSIT_BFX_File) if File.exist?(CSV_Test_File)
      @date_bfx_1 = Date.new(2017,1,10)
      @date_btx_1 = Date.new(2017,1,21)
      @date_btx_2 = Date.new(2017,1,10)
      VCR.use_cassette("rapflag", :record => :new_episodes) do
        expect(File.exist?(SUMMARY_DEPOSIT_BFX_File)).to eql(false)
        @deposit = RAPFLAG::History.new('deposit', 'BFX')
        @deposit.fetch_csv_history
        @deposit.create_summary
      end
    end
    it 'should have NOT generated a exchange BTC summary CSV file' do
      expect(File.exist?(SUMMARY_EXCHANGE_BTC_File)).to eql(false)
    end
    it 'should have NOT generated a correct summary CSV file' do
      expect(File.exist?(SUMMARY_DEPOSIT_BFX_File)).to eql(true)
      lines = IO.readlines(SUMMARY_DEPOSIT_BFX_File)
      expect(lines.first.chomp).to eql('currency,date,income,balance,rate,balance_in_usd')
      expect(lines[1].chomp).to eql('BFX,2016.01.15,0.0,8.99788147,,')
    end
  end
end


spec/bitfinex_spec.rb
# encoding: utf-8
require 'spec_helper'
require 'rapflag/bitfinex'
require 'vcr'
require 'date'

VCR.configure do |config|
  config.cassette_library_dir = "fixtures/vcr_cassettes"
  config.hook_into :faraday # or :fakeweb
end

VCR.eject_cassette # we use insert/eject around each example
def gen_deposit
  @date_bfx_1 = Date.new(2017,1,10)
  @date_btx_1 = Date.new(2017,1,21)
  @date_btx_2 = Date.new(2017,1,10)
  VCR.use_cassette("rapflag", :record => :new_episodes) do
    expect(File.exist?(BITFINEX_SUMMARY_DEPOSIT_BFX_File)).to eql(false)
    @deposit = RAPFLAG::Bitfinex.new('deposit', 'BFX')
    @deposit.fetch_csv_history
    @deposit.create_summary
  end
end
def gen_exchange
  @date_bfx_1 = Date.new(2017,1,10)
  @date_btx_1 = Date.new(2017,1,21)
  @date_btx_2 = Date.new(2017,1,10)
  VCR.use_cassette("rapflag", :record => :new_episodes) do
    expect(File.exist?(BITFINEX_SUMMARY_EXCHANGE_BTC_File)).to eql(false)
    @exchange = RAPFLAG::Bitfinex.new('exchange', 'BTC')
    @exchange.fetch_csv_history
    @exchange.create_summary
  end
end
def gen_dump(wallet = 'deposit')
  @date_bfx_1 = Date.new(2017,1,10)
  @date_btx_1 = Date.new(2017,1,21)
  @date_btx_2 = Date.new(2017,1,10)
  VCR.use_cassette("btc_#{wallet}", :record => :new_episodes) do
    @history_name = File.join(RAPFLAG.outputDir, 'bitfinex', "BTC_#{wallet}_history.csv")
    expect(File.exist?(@history_name)).to eql(false)
    @exchange = RAPFLAG::Bitfinex.new(wallet, 'BTC')
    @exchange.fetch_csv_history
    @exchange.dump_history
    @exchange.create_summary
    @exchange.create_total
  end
end
describe RAPFLAG::Bitfinex do
  OUTPUT_ROOT = File.expand_path(File.join(__FILE__, '..', 'output'))
  Bitfinex_CSV_Test_File             = File.join(RAPFLAG.outputDir, 'bitfinex', 'BTC_exchange.csv')
  BITFINEX_SUMMARY_DEPOSIT_BFX_File  = File.join(RAPFLAG.outputDir, 'bitfinex', 'BTC_deposit_summary.csv')
  BITFINEX_SUMMARY_EXCHANGE_BTC_File = File.join(RAPFLAG.outputDir, 'bitfinex', 'BTC_exchange_summary.csv')
  BITFINEX_SUMMARY_TRADING_BTC_File  = File.join(RAPFLAG.outputDir, 'bitfinex', 'BTC_trading_summary.csv')
  BITFINEX_TOTAL_BTC_File            = File.join(RAPFLAG.outputDir, 'bitfinex', 'BTC_total.csv')
  BITFINEX_SUMMARY_DEPOSIT_BFX_File  = File.join(RAPFLAG.outputDir, 'bitfinex', 'BFX_deposit_summary.csv')

  # include ServerMockHelper
  before(:all) do
  end
  after(:all) do
  end
  context 'bitfinex' do
    before(:all) do
      FileUtils.rm_rf(RAPFLAG.outputDir)
      VCR.use_cassette("rapflag", :record => :new_episodes) do
        expect(File.exist?(Bitfinex_CSV_Test_File)).to eql(false)
        @rap = RAPFLAG::Bitfinex.new('exchange', 'BTC')
        @rap.fetch_csv_history
        @rap.create_csv_file
      end
    end
    it 'should store into spec/output' do
      expect(OUTPUT_ROOT).to eq RAPFLAG.outputDir
    end
    context 'history' do
      it 'should have correct currency' do
        expect(@rap.currency).to eql('BTC')
      end
      it 'should have correct size' do
        expect(@rap.history.size).to eql(206)
      end
    end
  end
  context 'bitfinex CSV' do
    context 'csv' do
      it 'should have generated a correct CSV file' do
        expect(File.exist?(Bitfinex_CSV_Test_File)).to eql(true)
        lines = IO.readlines(Bitfinex_CSV_Test_File)
        expect(lines.first.chomp).to eql('currency,amount,balance,description,date_time')
        expect(lines[1].chomp).to eql(
          'BTC,-0.00000005,0.0,Transfer of 0.0 BTC from wallet Exchange to Deposit on wallet Exchange,2016.12.03 21:20:47')
      end
      it 'should have some lines inverted if the timestamp match' do
        expect(File.exist?(Bitfinex_CSV_Test_File)).to eql(true)
        lines = IO.readlines(Bitfinex_CSV_Test_File)
        first  = lines.index("BTC,-0.0170128,11.0283072,Trading fees for 8.5064 BTC @ 442.06 on BFX (0.2%) on wallet Exchange,2016.02.21 23:48:48\n")
        second = lines.index("BTC,-0.0004,11.04532,Trading fees for 0.2 BTC @ 442.1 on BFX (0.2%) on wallet Exchange,2016.02.21 23:48:48\n")
        expect(second + 1).to eq first
      end
    end
  end
  context 'exchange option --clean' do
    before(:all) do
      gen_exchange
    end
    it 'should have generated a correct exchange summary CSV file' do
      expect(File.exist?(BITFINEX_SUMMARY_EXCHANGE_BTC_File)).to eql(true)
      lines = IO.readlines(BITFINEX_SUMMARY_EXCHANGE_BTC_File)
      expect(lines.first.chomp).to eql('currency,date,income,balance')
      expect(lines[1].chomp).to eql('BTC,2016.01.15,"",0.0')
      expect(lines[-1].chomp).to eql('BTC,2016.12.03,"",0.0')
    end
    it 'should have a balance for for each day' do
      expect(File.exist?(BITFINEX_SUMMARY_EXCHANGE_BTC_File)).to eql(true)
      lines = IO.readlines(BITFINEX_SUMMARY_EXCHANGE_BTC_File)
      first_date = Date.strptime(lines[1].chomp.split(RAPFLAG::COLUMN_SEPARATOR)[1], '%Y.%m.%d')
      last_date = Date.strptime(lines[-1].chomp.split(RAPFLAG::COLUMN_SEPARATOR)[1], '%Y.%m.%d')
      expect((last_date >  first_date)).to be true
      nr_dates = 323
      expect((last_date -  first_date).to_i).to eql nr_dates
      (1..nr_dates).each do |j|
        wish_date = (first_date + j).strftime('%Y.%m.%d')
        expect(lines.find_all{|line| line.index(wish_date)}.size).to eql 1
      end
    end
    it 'should have NOT have generated a correct summary deposit BFX CSV file' do
      expect(File.exist?(BITFINEX_SUMMARY_DEPOSIT_BFX_File)).to eql(false)
    end
  end
  context 'deposit option --clean' do
    before(:all) do
      FileUtils.rm_rf(RAPFLAG.outputDir)
      gen_deposit
      VCR.use_cassette("rapflag", :record => :new_episodes) do
        @exchange = RAPFLAG::Bitfinex.new('exchange', 'BTC')
        @exchange.fetch_csv_history
        @exchange.create_summary
      end
      VCR.use_cassette("rapflag", :record => :new_episodes) do
        @trading = RAPFLAG::Bitfinex.new('trading', 'BTC')
        @trading.fetch_csv_history
        @trading.create_summary
      end
      @trading.create_total
    end
    it 'should have generated a exchange BTC summary CSV file' do
      expect(File.exist?(BITFINEX_SUMMARY_EXCHANGE_BTC_File)).to eql(true)
    end
    it 'should have generated a trading BTC summary CSV file' do
      expect(File.exist?(BITFINEX_SUMMARY_TRADING_BTC_File)).to eql(true)
    end
    it 'should have generated a deposit BTC summary CSV file' do
      expect(File.exist?(BITFINEX_SUMMARY_DEPOSIT_BFX_File)).to eql(true)
    end
    it 'should generate a correct summary CSV file' do
      expect(File.exist?(BITFINEX_SUMMARY_DEPOSIT_BFX_File)).to eql(true)
      lines = IO.readlines(BITFINEX_SUMMARY_DEPOSIT_BFX_File)
      expect(lines.first.chomp).to eql('currency,date,income,balance')
      expect(lines[1].chomp).to eql('BFX,2016.01.15,"",0.0')
    end
  end
  context 'option --dump' do
    before(:all) do
      FileUtils.rm_rf(RAPFLAG.outputDir)
      RAPFLAG::Wallets.each{ |wallet| gen_dump(wallet) }
      @stichtag = '2017.05.23'
    end
    it 'should have generated a correct deposit history CSV file' do
      history_name = File.join(RAPFLAG.outputDir, 'bitfinex', "BTC_deposit_history.csv")
      expect(File.exist?(history_name)).to eql(true)
      lines = IO.readlines(history_name)
      expect(lines.first.chomp).to eql('currency,amount,balance,description,timestamp,date')
      expect(lines[1].chomp).to eql('BTC,0.00000183,26.23947379,Margin Funding Payment on wallet Deposit,1500773485.0,2017.07.23')
      expect(lines[-1].chomp).to eql('BTC,2.0,2.0,Deposit (BITCOIN) #614485 on wallet Deposit,1446157557.0,2015.10.29')
      stich_item = lines.find{|x| /#{@stichtag}/.match(x)}.chomp
      expect(stich_item).to eql 'BTC,-1.10078339,41.64921661,Transfer of 1.1008 BTC from wallet Deposit to Exchange on wallet Deposit,1495556898.0,2017.05.23'
    end
    it 'should have generated a correct exchange history CSV file' do
      history_name = File.join(RAPFLAG.outputDir, 'bitfinex', "BTC_exchange_history.csv")
      expect(File.exist?(history_name)).to eql(true)
      lines = IO.readlines(history_name)
      expect(lines.first.chomp).to eql('currency,amount,balance,description,timestamp,date')
      expect(lines[1].chomp).to eql('BTC,-2.9,0.0,Transfer of 2.9 BTC from wallet Exchange to Deposit on wallet Exchange,1497288653.0,2017.06.12')
      expect(lines[-1].chomp).to eql('BTC,12.65,12.65,Transfer of 12.65 BTC from wallet Trading to Exchange on wallet Exchange,1452859181.0,2016.01.15')
      stich_item = lines.find{|x| /#{@stichtag}/.match(x)}.chomp
      expect(stich_item).to eql 'BTC,-0.98257955,0.0,Exchange 0.98257955 BTC for USD @ 2220.8 on wallet Exchange,1495557186.0,2017.05.23'
    end
    it 'should have generated a correct trading history CSV file' do
      history_name = File.join(RAPFLAG.outputDir, 'bitfinex', "BTC_trading_history.csv")
      expect(File.exist?(history_name)).to eql(true)
      lines = IO.readlines(history_name)
      expect(lines.first.chomp).to eql('currency,amount,balance,description,timestamp,date')
      expect(lines[1].chomp).to eql('BTC,-12.65,0.0,Transfer of 12.65 BTC from wallet Trading to Exchange on wallet Trading,1452859181.0,2016.01.15')
      expect(lines[-1].chomp).to eql('BTC,12.65,12.65,Transfer of 12.65 BTC from wallet Deposit to Trading on wallet Trading,1452859061.0,2016.01.15')
      expect(lines.find{|x| /#{@stichtag}/.match(x)}).to eql nil
    end
    it 'should have generated a correct total CSV file' do
      total_name = File.join(RAPFLAG.outputDir, 'bitfinex', "BTC_total.csv")
      expect(File.exist?(total_name)).to eql(true)
      lines = IO.readlines(total_name)
      expect(lines.first.chomp).to eql('currency,date,total_income,total_balance')
      expect(lines[1].chomp).to eql('BTC,2015.10.29,0.0,2.0')
      expect(lines[-1].chomp).to eql('BTC,2017.07.23,1.83e-06,26.23947379')

      # If no trades were given we must repeat the amount for each day
      first_date = lines.find_all{ |x| /BTC,2016.07.01/.match(x)}[0]
      amounts = /BTC,2016.07.01,(.*)/.match(first_date)[1]
      all_days = lines.find_all{ |x| x.index(amounts) }
      expect(all_days = lines.find_all{ |x| x.index(amounts) }.size).to be > 1
      stich_item = lines.find{|x| /#{@stichtag}/.match(x)}.chomp
      expect(stich_item).to eql 'BTC,2017.05.23,0.030693480000000002,41.64921661'
    end
  end
end


lib/rapflag/config.rb
require 'yaml'
require 'bitfinex-api-rb'

module RAPFLAG

  # Configure the client with the proper KEY/SECRET, you can create a new one from:
  # https://www.bitfinex.com/api
  config_file = nil
  files_to_test =  [ File.join(Dir.pwd, 'etc', 'config.yml'),
                     File.expand_path(File.join(File.dirname(__FILE__), '..', '..', 'etc', 'config.yml'))
                   ]
  files_to_test.each do |file_to_check|
    if File.exist?(file_to_check)
      config_file = file_to_check
      puts "Using config from #{file_to_check}"
      break
    end
  end
  if defined?(RSpec)
    Config = Hash.new
  else
    unless config_file && File.exist?(config_file)
      puts "You must first add a config.yml. Use one of the following places #{files_to_test}"
      exit 2
    end

    Config= YAML.load_file(config_file)
  end
  Config['currencies'] ||= ['BTC', 'BFX', 'XMR', 'ZEC', 'XRP', 'ETH', 'LTC' 'BCH']
  Config['currencies'] << 'USD' unless Config['currencies'].index('USD')
  Config['currencies'].sort!
end


lib/rapflag/history.rb
require 'csv'
require 'open-uri'
require 'faraday'
require 'fileutils'
require 'date'

module RAPFLAG
  class History
    @@global_totals ||= {}
    attr_reader :history, :wallet, :currency, :btc_to_usd, :bfx_to_usd
    DATE_FORMAT = '%Y.%m.%d'
    DATE_TIME_FORMAT = '%Y.%m.%d %H:%M:%S'
    INCOME_PATTERN = /Dividend|distribution|Settlement|Wire Withdrawal fee|Crypto Withdrawal fee|\
Bitcoin Gold snapshot step3 on Funding wallet|Trading fees for|Margin Funding Payment on wallet/i

    def initialize(wallet = 'trading', currency = 'USD')
      @wallet = wallet
      @currency = currency
    end

    def create_csv_file
      out_file = File.join(RAPFLAG.outputDir, "#{self.class.to_s.split('::').last.downcase}/#{@currency}_#{@wallet}.csv")
      FileUtils.makedirs(File.dirname(out_file))
      CSV.open(out_file,'w',
          :write_headers=> true,
          :col_sep => COLUMN_SEPARATOR,
          :headers => ['currency',
                      'amount',
                      'balance',
                      'description',
                      'date_time',
                      ] #< column header
        ) do |csv|
        skip_next = false
        @history.each_with_index do | hist_item, index|
          if skip_next
            skip_next = false
            next
          end
          timestamp = Time.at(hist_item['timestamp'].to_i).strftime(DATE_TIME_FORMAT)
          next_timestamp = @history[index+1] && Time.at(@history[index+1]['timestamp'].to_i).strftime(DATE_TIME_FORMAT)
          if next_timestamp.eql?(timestamp)
            next_item = @history[index + 1]
            this_day = Date.parse(timestamp).to_s
            prev_day = (Date.parse(timestamp)-1).to_s
            this_tx = history.find_all{|x| Time.at(x['timestamp'].to_i).to_date.to_s.eql?(this_day)}
            prev_tx = history.find{|x| Time.at(x['timestamp'].to_i).to_date < Date.parse(timestamp)}
            @total = prev_tx ?  prev_tx['balance'].to_f : 0.0
            this_tx.each{|x|to_add =  x['amount'].to_f; @total = @total + to_add}
            if @total == next_item['balance'].to_f
              second = hist_item
              first  = next_item
            else
              first  = hist_item
              second = next_item
            end
            skip_next = true
            csv << [ first['currency'],
                    first['amount'],
                    first['balance'],
                    first['description'],
                    Time.at(first['timestamp'].to_i).strftime(DATE_TIME_FORMAT),
                    ]
            csv << [ second['currency'],
                    second['amount'],
                    second['balance'],
                    second['description'],
                    Time.at(second['timestamp'].to_i).strftime(DATE_TIME_FORMAT),
                    ]
          else
            csv << [ hist_item['currency'],
                    hist_item['amount'],
                    hist_item['balance'],
                    hist_item['description'],
                    timestamp,
                    ]
          end
        end
      end

      sums = {}
      @history.each do | hist_item|
        key = /^[^\d]+/.match(hist_item['description'])[0].chomp
        value = hist_item['amount'].to_f
        if sums[key]
          sums[key] +=  value
        else
          sums[key]  =  value
        end
      end

      puts
      puts "Summary for #{@wallet} #{@currency} (#{@history.size} entries}"
      sums.each do |key, value|
        puts " #{sprintf('%40s', key)} is #{value}"
      end
    end

    Struct.new("Daily", :date, :amount, :balance, :description, :income)

    def create_summary
      @daily = {}
      out_file = File.join(RAPFLAG.outputDir, "#{self.class.to_s.split('::').last.downcase}/#{@currency}_#{@wallet}_summary.csv")
      FileUtils.makedirs(File.dirname(out_file))
      sorted_history =  @history.sort{|x,y| x['timestamp'] <=> y['timestamp']}
      sorted_history.each_with_index do | hist_item, index|
        date = Time.at(hist_item['timestamp'].to_i).strftime(DATE_FORMAT)
        next if @daily[date]
        @balance_last_day = (index == 0) ? 0.0 : sorted_history[index-1]['balance'].to_f
        this_day = Struct::Daily.new(date, 0.0, 0.0, hist_item['description'], 0.0)
        same_date = @history.select{|v| v['date'].eql?(date)}
        same_date.each do |info_item|
          amount = info_item['amount'].to_f
          info_item['time'] = Time.at(hist_item['timestamp'].to_i).utc
          this_day.amount += amount
          this_day.income += amount if INCOME_PATTERN.match(info_item['description'])
        end;
        @balance_last_day += this_day.amount
        this_day.balance = sorted_history[index+same_date.size-1]['balance'].to_f
        unless (diff = (@balance_last_day - sorted_history[index+same_date.size-1]['balance'].to_f)) < 0.01
          # puts "diff #{diff} for #{@currency}:#{index} #{date} #{same_date}"
          # Looks like the exchange rates varied between the day and therefore the calculated diff is not always correct
        end
        @daily[date] = this_day
      end
      previous_date = nil
      saved_info = nil
      CSV.open(out_file,'w',
          :write_headers=> true,
          :col_sep => COLUMN_SEPARATOR,
          :headers => ['currency',
                       'date',
                      'income',
                      'balance',
                      ] #< column header
        ) do |csv|

        @daily.each do |date, info|
          strings = date.split('.')
          fetch_date = Date.new(strings[0].to_i, strings[1].to_i, strings[2].to_i)
          (1..((fetch_date - previous_date).to_i.-1)).each do |j|
            intermediate = (previous_date + j).strftime(DATE_FORMAT)
            content = [@currency, intermediate, "",  saved_info.balance,]
            csv << content
            add_total(intermediate, 0.0, saved_info.balance)
          end if previous_date
          content = [@currency, date, info.income == 0.0 ? '': info.income, info.balance,]
          csv << content
          add_total(date, info.income, info.balance)
          previous_date = fetch_date
          saved_info = info
        end
      end
    end
    private
    def should_break(string_date)
      return false
      aDate = Date.parse(string_date)
      @currency.eql?('BTC') && aDate.year == 2017 && aDate.month == 05 && aDate.day == 23
    end
    def add_total(date, income, balance)
      if date.is_a?(String)
        key = [@currency, date]
        sdate = date
      elsif date.is_a?(Date)
        sdate = date.strftime(DATE_FORMAT)
        key = [@currency, sdate]
      end
      @@global_totals[key]    ||= [0, 0]
      if should_break(sdate)
        puts "add_total #{sdate} #{income} #{balance} to #{@@global_totals[key]} #{@wallet}"
        binding.pry
      end
      @@global_totals[key][0] += income
      @@global_totals[key][1] += balance
    end
    public
    def create_total
      this_currency = @@global_totals.select{|v, k| v[0].eql?(@currency)}
      day_index = 1
      sorted = Hash[ this_currency.sort_by { |key, val| key[day_index] } ]
      dates = sorted.keys.collect{|x| x[day_index]}.uniq
      total_filename = File.join(RAPFLAG.outputDir, "#{self.class.to_s.split('::').last.downcase}/#{@currency}_total.csv")
      FileUtils.makedirs(File.dirname(total_filename)) unless File.directory?(File.dirname(total_filename))
      total_file = CSV.open(total_filename, 'w+',
          :write_headers=> true,
          :col_sep => COLUMN_SEPARATOR,
          :headers => ['currency',
                       'date',
                      'total_income',
                      'total_balance',
                      ] #< column header
        )  do |csv|
        dates.each do |date|
          total_income = 0
          total_balance = 0
          this_day = this_currency.select{|v, k| v[day_index].eql?(date)}
          binding.pry if date && should_break(date)
          this_day.each do |key, info|
            total_income += info[0]
            total_balance += info[1]
          end
          csv << [@currency, date, total_income, total_balance]
        end
      end
    end
  end
end
# currency,date,income,balance
